lib/core/utils/preload.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * Validated the preload object
+ * @param {Object | boolean} preload configuration object or boolean passed via the options parameter to axe.run
+ * @return {boolean}
+ * @private
+ */
+function isPreloadValidObject(preload) {
+	return (
+		typeof preload === 'object' &&
+		preload.hasOwnProperty('assets') &&
+		Array.isArray(preload.assets) &&
+		preload.assets.length
+	);
+}
+
+/**
+ * Returns a boolean which decides if preload is configured
+ * @param {Object} options run configuration options (or defaults) passed via axe.run
+ * @return {boolean}
+ */
+function shouldPreload(options) {
+	if (!options) {
+		return false;
+	}
+
+	if (!options.preload) {
+		return false;
+	}
+
+	if (typeof options.preload === typeof true) {
+		return options.preload;
+	}
+
+	if (isPreloadValidObject(options.preload)) {
+		return true;
+	}
+
+	return false;
+}
+axe.utils.shouldPreload = shouldPreload;
+
+/**
+ * Constructs a configuration object representing the preload requested assets & timeout
+ * @param {Object} options run configuration options (or defaults) passed via axe.run
+ * @return {Object}
+ */
+function getPreloadConfig(options) {
+	// default fallback configuration
+	let out = {
+		assets: axe.constants.preloadAssets,
+		timeout: axe.constants.preloadAssetsTimeout
+	};
+
+	// if type is boolean
+	if (typeof options.preload === typeof true) {
+		return out;
+	}
+
+	// if type is object - ensure an array of assets to load is specified
+	if (isPreloadValidObject(options.preload)) {
+		const requestedAssets = [];
+		options.preload.assets.forEach(asset => {
```

Comment:
Looks like we're constructing a new array of assets in this loop. Maybe we should `.map()` and `.filter()` instead of this `.forEach()`?

IMO the usage of `.map()` makes the intent of the code a little easier to pickup.

Reply:
Came up with a different workflow, so can throw earlier, than in an iteration, also keeps the line of sight in the code to the left.